sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,652 @@
+    private final String query;
+    @Nullable
+    private final String username;
+    @Nullable
+    private final String password;
+
+    public BoundedJdbcSource(DataSource dataSource, String query, String username,
+                             String password) {
+      this.dataSource = dataSource;
+      this.query = query;
+      this.username = username;
+      this.password = password;
+    }
+
+    @Override
+    public Coder getDefaultOutputCoder() {
+      return SerializableCoder.of(JdbcDataRecord.class);
+    }
+
+    @Override
+    public void validate() {
+      Preconditions.checkNotNull(dataSource, "dataSource");
+      Preconditions.checkNotNull(query, "query");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      builder.add(DisplayData.item("dataSource", dataSource.getClass().getName()));
+      builder.add(DisplayData.item("query", query));
+      builder.addIfNotNull(DisplayData.item("username", username));
+    }
+
+    @Override
+    public boolean producesSortedKeys(PipelineOptions options) {
+      return false;
+    }
+
+    @Override
+    public BoundedReader createReader(PipelineOptions options) {
+      return new BoundedJdbcReader(this);
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions options) {
+      // there's no actual way to get the estimated size of a query without performing the
+      // query. So it would be a bottleneck in the JdbcIO.
+      return 0L;
+    }
+
+    @Override
+    public List<BoundedSource<JdbcDataRecord>> splitIntoBundles(long desiredBundleSizeBytes,
+                                                PipelineOptions options) {
+      // related to getEstimatedSizeBytes, the only way to define the split is by hand (user
+      // defined). We return an unique source for JDBC.
+      List<BoundedSource<JdbcDataRecord>> sources = new ArrayList();
+      sources.add(this);
+      return sources;
+    }
+
+  }
+
+  private static class BoundedJdbcReader extends BoundedSource.BoundedReader {
+
+    private final BoundedJdbcSource source;
+
+    private Connection connection;
+    private PreparedStatement statement;
+    private ResultSet resultSet;
+    private ResultSetMetaData resultSetMetaData;
+
+    private JdbcDataRecord current;
+
+    public BoundedJdbcReader(BoundedJdbcSource source) {
+      this.source = source;
+    }
+
+    @Override
+    public boolean start() {
+      try {
+        if (source.username != null) {
+          connection = source.dataSource.getConnection(source.username, source.password);
+        } else {
+          connection = source.dataSource.getConnection();
+        }
+        statement = connection.prepareCall(source.query);
+        resultSet = statement.executeQuery();
+        resultSetMetaData = resultSet.getMetaData();
+
+        return advance();
+      } catch (Exception e) {
+        LOGGER.error("Can't connect to the database", e);
+        return false;
```

Comment:
From the point of view of the runner, this is simply a "return false" - meaning, the runner will happily declare that the bundle is complete because we reached EOF. From the point of view of the user, this is silent data loss signaled only by an error message in worker logs. I don't think that's what you intended, so it's better to rethrow the exception, both here and in advance().